src/v/kafka/server/group_manager.cc

```diff
@@ -605,6 +620,96 @@ group_manager::txn_offset_commit(txn_offset_commit_request&& r) {
       });
 }
 
+ss::future<cluster::begin_group_tx_reply>
+group_manager::begin_tx(cluster::begin_group_tx_request&& r) {
+    auto catchup_lock = get_catchup_lock(r.ntp);
+    if (!catchup_lock || catchup_lock->try_read_lock()) {
+        // transaction operations can't run in parallel with loading
+        // state from the log (happens once per term change)
+        vlog(klog.trace, "can't process a tx: coordinator_load_in_progress");
+        return ss::make_ready_future<cluster::begin_group_tx_reply>(
+          make_begin_tx_reply(cluster::tx_errc::coordinator_load_in_progress));
+    }
+
+    return do_begin_tx(std::move(r)).finally([catchup_lock] {
+        catchup_lock->read_unlock();
+    });
+}
+
+ss::future<cluster::begin_group_tx_reply>
+group_manager::do_begin_tx(cluster::begin_group_tx_request&& r) {
+    auto error = validate_group_status(
+      r.ntp, r.group_id, offset_commit_api::key);
+    if (error != error_code::none) {
+        if (error == error_code::not_coordinator) {
+            co_return make_begin_tx_reply(cluster::tx_errc::not_coordinator);
+        } else {
+            co_return make_begin_tx_reply(cluster::tx_errc::timeout);
+        }
+    }
+
+    auto partition = _partitions.find(r.ntp)->second;
```

Comment:
what guarantees that `_partitions.find(r.ntp)` != `_partitions.end()`?

Reply:
nothing, fix by checking it